packages/supplementary-contracts/contracts/TokenUnlocking.sol

```diff
@@ -1,3 +1,332 @@
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/// @title TokenUnlocking
+/// @notice Contract for managing Taiko token unlocking.
+///
+/// It manages only unlocking and vested tokens will be deposited into this contract (through
+/// 'depositToGrantee()' function) when the purchase notice sent out by Taiko, is paid. Unlocking
+/// will be a 4-year immutable period, with a 1-year cliff, counting from TGE.
+///
+/// Vesting will be a regular (quarterly / twice a year, TBD) 'off-chain' legal payment action,
+/// where those purschase notices can be exercised (and paid up-front, so before depoists made).
+/// Once tokens are deposited into this contract it cannot be forfeited anymore. If an
+/// employment is ended, the company (Taiko) simply does not send out purchase notices anymore, so
+/// that more tokens would not be deposited, beside the eligible proportion of that same vesting
+/// release period (e.g.: if Bob spent 1 month at Taiko out of that quarterly/half-yearly vesting,
+/// those will be deposited of course).
+///
+/// We should deploy multiple instances of this contract per grantee and per grant. So each person
+/// should have the same amount of contract deployed as grants granted (grant 1, grant 2, etc.)
+/// @custom:security-contact security@taiko.xyz
+contract TokenUnlocking is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+
+    // Variables, which shall be same for everybody, it is part of the GrantInfo struct.
+    struct UnlockTiming {
+        /// @notice Shall be TGE date.
+        uint64 unlockStartDate;
+        /// @notice The end date of cliff period. Shall be 1 year post TGE (date).
+        uint64 unlockCliffDate;
+        /// @notice Shall be 4 years (period)
+        uint32 unlockPeriod;
+    }
+
+    // Variables, which shall be unique per grantee
+    struct GrantInfo {
+        // This variable has to be set, before we handle any withdrawals
+        UnlockTiming unlockTiming;
+        /// @notice It is basically the same as "amount deposited" or "total amount vested" (so
+        /// far).
+        uint128 amountVested;
+        /// @notice Represents how many tokens withdrawn already and helps with: withdrawable
+        /// amount.
+        // - The current "withdrawable amount" is determined by the help of this variable =
+        // (amountVested *(% of unlocked) ) - amountWithdrawn
+        uint128 amountWithdrawn;
+        /// @notice The address of the recipient.
+        address grantRecipient;
+        // This can define which grant is this (grant 1, grant 2, etc.). Only informative
+        // purposes.
+        uint8 grantId;
+    }
+
+    // Holding any grant information
+    GrantInfo public grantInfo;
+
+    /// @notice The Taiko token address.
+    address public taikoToken;
+
+    /// @notice The shared vault address, from which tko token deposits will be triggered by the
+    /// depositToGrantee() function
+    address public sharedVault;
```

Comment:
Since the grantee has paid, lets transfer the TKO token directly from "msg.sender" to this contract without holding purchased tokens in the shared vault.

Reply:
We still need a "trigger deposit" and that deposit shall be coming from a vault but thru this contract. 
Otherwise we cannot keep track of deposited ERC20, since there is no fallback or receive() funciton which is called. So let's say TaikoLabs can say, it deposited 100K tokens, but indeed it did not.
All i'm saying is: this `depositToGrantee` should trigger a deposit into this contract (so that we can keep track of the indeed vested amount).